# PROJECT
## Applications of Big Data

### 0. Importing libraries

In [1]:
import pandas as pd

### 1. Data processing

#### 1.0. Data loading

In [2]:
products = pd.read_csv('dataset/products.csv')
products.head()

id                                               name   price  \
0   74021317  [Nhập 1212A10K giảm 10K đơn 199K] Cây Cam Ngọt...   68040   
1  249009518  Sổ kế hoạch planner 18 tháng 2023-2024 SDstati...   12000   
2  273612826  Tạp chí Pi- Hội Toán học Việt Nam số 12/ tháng...   35000   
3  273536931  Kinh Doanh Thời Trang - 8 Bước xây dựng, vận h...  169000   
4  182784585  (Tặng Poster 10 Kĩ năng sơ cấp cứu cơ bản) Sơ ...  240000   

   original_price  discount_rate  quantity_sold  rating_average  review_count  \
0          108000             37          70416             5.0         13151   
1           39000             69             32             0.0             0   
2           35000              0              0             0.0             0   
3          199000             15             49             5.0             4   
4          240000              0             58             5.0             8   

   seller_id             category       spid  \
0          1          Tiểu Thuyết   74021318   
1        148      Sổ Tay Các Loại  249009520   
2        469  Tạp Chí - Catalogue  273612827   
3       2953     Sách kinh tế học  273536932   
4      52987           Sách Y Học  182784586   

                           seller_name                author_name  
0                         Tiki Trading  José Mauro de Vasconcelos  
1                         SDstationery                        NaN  
2                           Tạp Chí Pi              nhieu tac gia  
3                  GIVER BOOKS & MEDIA                        NaN  
4  Tổ chức Giáo dục sức khỏe Wellbeing                        NaN

In [3]:
reviews = pd.read_csv('dataset/reviews.csv')
reviews.head()

id  created_at  rating            title  \
0  12559756  1633960872       5  Cực kì hài lòng   
1  16979365  1657206097       5  Cực kì hài lòng   
2  14069617  1640442831       5  Cực kì hài lòng   
3  12322259  1632568671       5  Cực kì hài lòng   
4  18368714  1670438913       5  Cực kì hài lòng   

                                             content  thank_count  \
0  Có những người bước đến, họ lấp đầy hạnh phúc ...          389   
1  Thấy nhiều bạn chê tiki gói hàng quá, may sao ...           14   
2  Bìa cực xinh, tiki giao hàng nhanh, sách không...           26   
3  Nội dung của sách thì không phải bàn đến rồi, ...           16   
4  Một cuốn sách rất đáng đọc về tình yêu thương ...            1   

        customer_name  customer_id  product_id  seller_product_id  
0             Vân Anh     22051463    74021317           74021318  
1  Phan Thảo Yến Nhím     27791831    74021317           74021318  
2      Trần Thị Trang     17748750    74021317           74021318  
3            Khánh Ly     10149686    74021317           74021318  
4         Phương Linh     28545286    74021317           74021318

In [4]:
print(products.shape)
print(reviews.shape)

(2030, 13)
(605259, 10)


**Dữ liệu được thu thập về từ `tiki`, mô tả về tất cả các sách trong mục `nhà sách tiki`. Gồm hai bộ dữ liệu: `products` và `reviews`.**

*`products` gồm các thông tin về sản phẩm sách trong nhà sách tiki.*
- Dữ liệu có 2030 dòng và 13 cột. Trong đó:
    - `id`: mã sản phẩm
    - `name`: tên sản phẩm
    - `price`: giá sản phẩm
    - `original_price`: giá niêm yết
    - `discount_rate`: tỉ lệ giảm giá
    - `quantity_sold`: số lượng đã bán
    - `rating_average`: đánh giá trung bình
    - `review_count`: số lượng đánh giá
    - `seller_id`: mã người bán
    - `category`: danh mục/thể loại
    - `seller_name`: tên người bán
    - `author_name`: tác giả
    - `spid`: mã sản phẩm

*`reviews` gồm các đánh giá về sản phẩm được đề cập trong products.*
- Dữ liệu có 605259 dòng và 10 cột. Trong đó:
    - `id`: mã đánh giá
    - `product_id`: mã sản phẩm
    - `rating`: đánh giá
    - `content`: nội dung đánh giá
    - `author_name`: tên tác giả
    - `title`: tiêu đề
    - `created_at`: thời gian đánh giá
    - `customer_id`: mã khách hàng
    - `customer_name`: tên khách hàng
    - `thank_count`: số lượng lượt thích đánh giá
    - `seller_product_id`: mã sản phẩm của người bán

#### 1.1. Duplicated

- Xóa cột `spid` vì nó trùng với cột `id`. (`products`)

In [5]:
products = products.drop(['spid'], axis=1)

In [6]:
products.duplicated().sum()

21

Qua kiểm tra thì có `21` dòng trùng trong tập products. Các dòng này bị trùng trong quá trình thu thập dữ liệu nên ta sẽ xóa đi.

In [7]:
#products[products['id'].duplicated(keep=False)].sort_values(by='id')
products = products.drop_duplicates()

Ta sẽ làm tương tự với tập dữ liệu `reviews`.

Xóa cột `seller_product_id` vì nó trùng với cột `product_id`.

In [8]:
reviews = reviews.drop(['seller_product_id'], axis=1)

In [9]:
reviews.duplicated().sum()

14774

Các dòng này bị trùng trong quá trình thu thập dữ liệu và nó chỉ chiếm khoảng `2,4%` dữ liệu nên ta sẽ xóa đi.

In [10]:
reviews = reviews.drop_duplicates()

Tuy nhiên qua quan sát nhóm thấy ở tập dữ liệu `reviews` trùng nhau ở cột `id` mà nó chỉ khác nhau ở 1 cột. Điều này có thể là do lỗi trong quá trình thu thập dữ liệu nên nhóm sẽ xóa đi các dòng trùng nhau ở tập `reviews`.

In [11]:
reviews[reviews['id'].duplicated(keep=False)].sort_values(by='id')

id  created_at  rating            title content  thank_count  \
561250  14779771  1643853610       5  Cực kì hài lòng     NaN            3   
561252  14779771  1643853610       5  Cực kì hài lòng     NaN            2   
529064  16392632  1653012913       5  Cực kì hài lòng     NaN            0   
529065  16392632  1653012913       5  Cực kì hài lòng     NaN            0   

                customer_name  customer_id  product_id  
561250            Nguyễn Long     12536047   146223395  
561252            Nguyễn Long     12536047   146223395  
529064  Trần Dương Minh Quang      7603930    67991600  
529065             Khách Hàng      7603930    67991600

Xóa 2 hằng có `index` là `561252` và `529065` vì chúng bị trùng với các dòng khác.

In [12]:
reviews = reviews.drop([56152, 529065])

#### 1.2. Missing values

In [13]:
products.isnull().sum()

id                  0
name                0
price               0
original_price      0
discount_rate       0
quantity_sold       0
rating_average      0
review_count        0
seller_id           0
category            0
seller_name         2
author_name       374
dtype: int64

Cột `seller_name` bị thiếu có thể là do các sản phẩm này không có tên người bán hoặc do lỗi trong quá trình thu thập dữ liệu. Qua kiểm tra thì nhóm phát hiện 1 giá trị bị thiếu ở cột `seller_name` có giá trị `seller_id` của 1 người bán. Nhóm sẽ thay thế giá trị này bằng tên người bán tương ứng. Còn giá trị thiếu còn lại thì không trùng `seller_id` nên nhóm sẽ thay thế bằng `seller_id` tướng ứng.

In [14]:
products['seller_name'] = products.apply(lambda row: products[products['seller_id']==row['seller_id']]['seller_name'].iloc[0] 
                                         if pd.isnull(row['seller_name']) else row['seller_name'], axis=1)

In [15]:
products['seller_name'].fillna(products['seller_id'], inplace=True)

Qua kiểm tra các giá trị thiếu của cột `author_name` là do các sản phẩm không có tên tác giả. Ta sẽ thay thế các giá trị này bằng `Unknown`.

In [16]:
products['author_name'].fillna('Unknown', inplace=True)

In [17]:
products.isnull().sum()

id                0
name              0
price             0
original_price    0
discount_rate     0
quantity_sold     0
rating_average    0
review_count      0
seller_id         0
category          0
seller_name       0
author_name       0
dtype: int64

In [18]:
reviews.isnull().sum()

id                    0
created_at            0
rating                0
title                 5
content          392619
thank_count           0
customer_name      6879
customer_id           0
product_id            0
dtype: int64

In [19]:
reviews[((reviews['rating']==5) )]

id  created_at  rating            title  \
0       12559756  1633960872       5  Cực kì hài lòng   
1       16979365  1657206097       5  Cực kì hài lòng   
2       14069617  1640442831       5  Cực kì hài lòng   
3       12322259  1632568671       5  Cực kì hài lòng   
4       18368714  1670438913       5  Cực kì hài lòng   
...          ...         ...     ...              ...   
605249   9561654  1620393883       5  Cực kì hài lòng   
605250   9554276  1620370994       5  Cực kì hài lòng   
605251   9022442  1618119132       5  Cực kì hài lòng   
605252   8726727  1616919902       5  Cực kì hài lòng   
605253   7517182  1612414399       5  Cực kì hài lòng   

                                                  content  thank_count  \
0       Có những người bước đến, họ lấp đầy hạnh phúc ...          389   
1       Thấy nhiều bạn chê tiki gói hàng quá, may sao ...           14   
2       Bìa cực xinh, tiki giao hàng nhanh, sách không...           26   
3       Nội dung của sách thì không phải bàn đến rồi, ...           16   
4       Một cuốn sách rất đáng đọc về tình yêu thương ...            1   
...                                                   ...          ...   
605249                                                NaN            0   
605250                                                NaN            0   
605251                                                NaN            0   
605252                                                NaN            0   
605253                                                NaN            0   

             customer_name  customer_id  product_id  
0                  Vân Anh     22051463    74021317  
1       Phan Thảo Yến Nhím     27791831    74021317  
2           Trần Thị Trang     17748750    74021317  
3                 Khánh Ly     10149686    74021317  
4              Phương Linh     28545286    74021317  
...                    ...          ...         ...  
605249           HaThuy Vu     10750423    75996803  
605250     Luu Quynh Huong     14854794    75996803  
605251             vo nhan      6698905    75996803  
605252           Xuân Xuân      5668673    75996803  
605253            Tô Nhung        11878    75996803  

[518401 rows x 9 columns]

Ở cột `title` bị thiếu có thể là do người dùng không muốn để lại bình luận. Qua quan sát thì cát giá trị này điều có `rating` là `5`. Ngoài ra thì các có dòng có `rating` là `5` thì `title` không bị thiếu đa phần là `Cực kỳ hài lòng` còn lại là `Hài lòng`, ... Nên nhóm sẽ thay thế các giá trị thiếu bằng `Cực kỳ hài lòng`.

In [20]:
reviews['title'] = reviews['title'].fillna('Cực kỳ hài lòng')

Ở cột `customer_name` nhóm sẽ thay thế các giá trị thiếu bằng những `customer_name` có giá tương ứng dựa vào `customer_id`. Tuy nhiên cũng có một số `customer_id` không có `customer_name` tương ứng. Nhóm sẽ thay thế các giá trị này bằng giá trị của `customer_id`.

In [21]:
def fill_na(row):
    if pd.isnull(row['customer_name']):
        rows = reviews[(reviews['customer_id'] == row['customer_id']) & (reviews['customer_name'].notnull())]
        if not rows.empty:
            return rows['customer_name'].iloc[0]
        else:
            return row['customer_id']
    else:
        return row['customer_name']

In [22]:
reviews['customer_name'] = reviews.apply(fill_na, axis=1)

In [23]:
reviews['customer_name'].fillna(reviews['customer_id'], inplace=True)

Ở cột `content` bị thiếu có thể là do người dùng không muốn để lại bình luận. Ta sẽ thay thế các giá trị này bằng `No comment`.

In [24]:
reviews['content'].fillna('No comment', inplace=True)

In [25]:
reviews.isnull().sum()

id               0
created_at       0
rating           0
title            0
content          0
thank_count      0
customer_name    0
customer_id      0
product_id       0
dtype: int64

#### 1.3. Data type

In [26]:
def open_object_dtype(s):
    dtypes = set(s.apply(lambda item : type(item)))
    return dtypes

products.apply(open_object_dtype)

id                               {<class 'int'>}
name                             {<class 'str'>}
price                            {<class 'int'>}
original_price                   {<class 'int'>}
discount_rate                    {<class 'int'>}
quantity_sold                    {<class 'int'>}
rating_average                 {<class 'float'>}
review_count                     {<class 'int'>}
seller_id                        {<class 'int'>}
category                         {<class 'str'>}
seller_name       {<class 'str'>, <class 'int'>}
author_name                      {<class 'str'>}
dtype: object

Chuyển kiểu dữ liệu cột `seller_name` sang kiểu `str` và đổi tên cột `id` --> thành `product_id`.

In [27]:
products['seller_name'] = products['seller_name'].astype('str')

In [28]:
products = products.rename(columns={'id':'product_id'})

Chuyển dữ liệu của các cột categorical sang viết thường, xóa khoảng trắng ở đầu và cuối, xoá các dấu câu, các kí tự đặc biệt và xóa khoảng trắng dư thừa.

In [29]:
products['category'] = products['category'].str.replace('- ', '')
products['category'] = products['category'].str.replace(',', '')
products['category'] = products['category'].str.lower().str.strip()

Ngoài ra nhóm còn phát hiện các lỗi chính tả khác trong cột `author_name`. Nhóm sẽ thay thế các giá trị này bằng các giá trị đúng.

In [30]:
products['author_name'] = products['author_name'].str.replace('Choi Kwanghuyn', 'Choi Kwanghyun')
products['author_name'] = products['author_name'].str.replace('Song Hong Binh', 'Song Hong Bing')
products['author_name'] = products['author_name'].str.replace('Rhowa Byrne', 'Rhonda Byrne')
products['author_name'] = products['author_name'].str.replace('Khailed Hosseini', 'Khaled Hosseini')
products['author_name'] = products['author_name'].str.replace('nhieu tac gia', 'Nhiều tác giả')

In [31]:
products['author_name'] = products['author_name'].str.replace('-', ' ')
products['author_name'] = products['author_name'].str.replace('.', '')

C:\Users\HoaTX\AppData\Local\Temp\ipykernel_4684\3755146916.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  products['author_name'] = products['author_name'].str.replace('.', '')


In [32]:
products['author_name'] = products['author_name'].str.replace('Baird TSpalding', 'Baird T Spalding')
products['author_name'] = products['author_name'].str.replace('Briad LWeiss', 'Briad L Weiss')
products['author_name'] = products['author_name'].str.replace('Robert TKiyosaki', 'Robert T Kiyosaki')
products['author_name'] = products['author_name'].str.replace('Stephen MRCovey', 'Stephen RCovey')
products['author_name'] = products['author_name'].str.lower().str.strip()

In [33]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 590483 entries, 0 to 605258
Data columns (total 9 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   id             590483 non-null  int64 
 1   created_at     590483 non-null  int64 
 2   rating         590483 non-null  int64 
 3   title          590483 non-null  object
 4   content        590483 non-null  object
 5   thank_count    590483 non-null  int64 
 6   customer_name  590483 non-null  object
 7   customer_id    590483 non-null  int64 
 8   product_id     590483 non-null  int64 
dtypes: int64(6), object(3)
memory usage: 45.1+ MB


Chuyển kiểu dữ liệu cột `customer_name` sang kiểu `str`.

In [34]:
reviews['customer_name'] = reviews['customer_name'].astype('str')

Chuyển kiểu dữ liệu cột `created_at` sang kiểu `datetime` và thêm các cột `year`, `month`, `day`, `weekend`, `hour` để phân tích dữ liệu.

In [35]:
reviews['created_at'] = pd.to_datetime(reviews['created_at'], unit='s')

In [36]:
reviews['year'] = reviews['created_at'].dt.year
reviews['month'] = reviews['created_at'].dt.month
reviews['day'] = reviews['created_at'].dt.day
reviews['hour'] = reviews['created_at'].dt.hour
reviews['weekday'] = reviews['created_at'].dt.weekday

In [37]:
#weekday = {0: 'Monday', 1: 'Tuesday', 2: 'Wednesday', 3: 'Thursday', 4: 'Friday', 5: 'Saturday', 6: 'Sunday'}
#reviews['weekday'] = reviews['weekday'].replace(weekday)

#### 1.4. Outliers

Qua kiểm tra thì dữ liệu không có Outlier.

#### 1.5. Save data

In [38]:
print(products.shape)
print(reviews.shape)

(2009, 12)
(590483, 14)


Sau quá trình xử lý dữ liệu, `products` còn `2009 dòng` và `12 cột` và `reviews` còn `590483 dòng` và `14 cột`.

In [39]:
df = pd.merge(reviews, products, on='product_id')

**Câu lệnh bên dưới chỉ thực hiện một lần duy nhất trên máy một thành viên duy nhất. Sau khi thực hiện, tiến hành lưu trữ df thành file csv, các thành viên khác chỉ cần load file này vào là được, không cần chạy lại câu lệnh `to_csv` để tránh mất thời gian.**

In [40]:
products.to_csv('dataset/products_clean.csv', index=False)
reviews.to_csv('dataset/reviews_clean.csv', index=False)

In [41]:
df.to_csv('dataset/merged_data.csv', index=False)